x/logic/predicate/bank.go

```diff
@@ -98,79 +80,89 @@
 		if err != nil {
 			return engine.Error(err)
 		}
-		bankKeeper := sdkContext.Value(types.BankKeeperContextKey).(types.BankKeeper)
 
-		bech32Addr, err := getBech32(env, account)
+		switch acc := env.Resolve(address).(type) {
+		case engine.Atom:
+			return engine.Delay(
+				func(ctx context.Context) *engine.Promise {
+					addr, err := sdk.AccAddressFromBech32(acc.String())
+					if err != nil {
+						return engine.Error(prolog.WithError(
+							engine.DomainError(prolog.ValidEncoding("bech32"), engine.NewAtom(acc.String()), env), err, env))
+					}
+					if exists := authKeeper.GetAccount(ctx, addr) != nil; !exists {
+						return engine.Bool(false)
+					}
+
+					return cont(env)
+				})
+		case engine.Variable:
+			return engine.DelaySeq(IterMap(Accounts(ctx, authQueryService),
+				func(it lo.Tuple2[sdk.AccountI, error]) engine.PromiseFunc {
+					return func(_ context.Context) *engine.Promise {
+						addr, err := lo.Unpack2(it)
+						if err != nil {
+							return engine.Error(prolog.WithError(engine.ResourceError(prolog.ResourceModule("auth"), env), err, env))
+						}
+						return engine.Unify(vm, address, engine.NewAtom(addr.GetAddress().String()), cont, env)
+					}
+				}))
+		default:
+			return engine.Error(engine.TypeError(prolog.AtomTypeAtom, address, env))
+		}
+	})
+}
+
+// fetchBalances is a helper function to fetch the balances of the given account using a given function which returns the coins for
+// the given address.
+func fetchBalances(vm *engine.VM, address engine.Term, balances engine.Term, coinsFn func(ctx context.Context,
+	bankKeeper types.BankKeeper, address sdk.AccAddress) sdk.Coins, cont engine.Cont, env *engine.Env,
+) *engine.Promise {
+	return engine.Delay(func(ctx context.Context) *engine.Promise {
+		bankKeeper, err := prolog.ContextValue[types.BankKeeper](ctx, types.BankKeeperContextKey, env)
 		if err != nil {
 			return engine.Error(err)
 		}
 
-		if bech32Addr != nil {
-			fetchedBalances := coinsFn(sdkContext, bankKeeper, nil, bech32Addr)
-			return engine.Unify(vm, CoinsToTerm(fetchedBalances), balances, cont, env)
-		}
+		return account(vm, address, func(env *engine.Env) *engine.Promise {
+			switch acc := env.Resolve(address).(type) {
+			case engine.Atom:
+				coins := coinsFn(ctx, bankKeeper, sdk.MustAccAddressFromBech32(acc.String()))
 
-		allBalances := bankKeeper.GetAccountsBalances(sdkContext)
-		promises := make([]func(ctx context.Context) *engine.Promise, 0, len(allBalances))
-		for _, balance := range allBalances {
-			address := balance.Address
-			bech32Addr, err = sdk.AccAddressFromBech32(address)
-			if err != nil {
-				return engine.Error(prolog.WithError(engine.ResourceError(prolog.ResourceModule("bank"), env), err, env))
+				return engine.Unify(vm, CoinsToTerm(coins), balances, cont, env)
+			default:
+				return engine.Error(engine.TypeError(prolog.AtomTypeAtom, address, env))
```

Comment:
The tests don’t cover the scenario where the address isn’t an `atom` because the continuation function is invoked by the `account` predicate, which guarantees that the `address` is always an `atom`. It’s not feasible to write a test for this case unless we _mock_ the `account` predicate to behave incorrectly, but honestly, that feels excessive.

Reply:
Also, keep in mind that we’re using CI with test coverage enabled, so there’s no need for you to check it.